### DISCUSSION TITLE
Investigate signal diffusion acceleration tools

### DISCUSSION THREAD

**Discussion by alice-i-cecile** [Leafwing-Studios/Emergence#288] [Ideas]
PDE approximation with operator learning: https://www.quantamagazine.org/new-neural-networks-solve-hardest-equations-faster-than-ever-20210419/

**Comment by alice-i-cecile** [comment]
Use fluid simulator? https://www.youtube.com/watch?v=Yzr9va5UtiE&t=4s

**Comment by alice-i-cecile** [comment]
More ideas:
- GPU acceleration
- Infrequent ticks
- Async tasks that are picked up opportunistically when spare threads are idle, especially in combination with frame pacing

**Comment by TimJentzsch** [comment]
What about a simple blurring technique? We could store the signal strengths in an image texture and then just blur it in a shader every X seconds.
We'd have to see which blurring technique to use for it to make sense with hex coordinates, but I assume that could be performant enough.

See also [this video](https://youtu.be/X-iSQQgOd1A?t=776) where a technique like this was used for slime simulation.

**Comment by alice-i-cecile** [comment]
For blurring, I would be worried that the signals wouldn't respect walls or other obstructions, which leads to very poor pathfinding behavior.

**Comment by bzm3r** [comment]
> What about a simple blurring technique? We could store the signal strengths in an image texture and then just blur it in a shader every X seconds. We'd have to see which blurring technique to use for it to make sense with hex coordinates, but I assume that could be performant enough.
> 
> See also [this video](https://youtu.be/X-iSQQgOd1A?t=776) where a technique like this was used for slime simulation.

How would we get the data back from the GPU? Typically, reading data back from the GPU is not cheap.

**Comment by TimJentzsch** [comment]
> > What about a simple blurring technique? We could store the signal strengths in an image texture and then just blur it in a shader every X seconds. We'd have to see which blurring technique to use for it to make sense with hex coordinates, but I assume that could be performant enough.
> > See also [this video](https://youtu.be/X-iSQQgOd1A?t=776) where a technique like this was used for slime simulation.
> 
> How would we get the data back from the GPU? Typically, reading data back from the GPU is not cheap.

We'd have to send the texture to the GPU, process it there and then send the modified texture back to the CPU.
I assume at a certain size this will be faster because of the faster processing time, but we'd need to benchmark it to know for sure

**Comment by JasmineLowen** [comment]
A friend of mine just showed me [this](https://gdcvault.blazestreaming.com/?id=da7c7939ec364ad5a0c75fb52b306f91&videoid=a1df8d31e7928a2e2d3a1025243) and it kind of reminded me of the problem outlined here. Maybe it helps
